File: include/nole.py
import sys, os, time, re, unicodedata

class NolError(Exception):
    def __init__(self, msg, line=None, col=None):
        super().__init__(f"{msg} at {line}:{col}" if line else msg)
        self.line, self.col = line, col

class NolParser:
    def __init__(self, text, nole=False):
        self.text = text; self.pos = 0; self.line = 1; self.col = 1
        self.start_time = time.time(); self.depth = 0; self.nole = nole
    def peek(self, n=0): return self.text[self.pos + n] if self.pos + n < len(self.text) else None
    def advance(self):
        c = self.peek(); self.pos += 1
        if c == '\n': self.line += 1; self.col = 1
        else: self.col += 1
        return c
    def skip_ws(self):
        while self.pos < len(self.text):
            if self.text[self.pos].isspace(): self.advance()
            elif self.text[self.pos] == '#':
                self.advance()
                if self.peek() == '#':
                    self.advance()
                    while self.pos < len(self.text) - 1 and not (self.text[self.pos] == '#' and self.text[self.pos+1] == '#'): self.advance()
                    if self.pos < len(self.text): self.advance(); self.advance()
                else:
                    while self.pos < len(self.text) and self.text[self.pos] != '\n': self.advance()
            else: break
    def read_str(self, q):
        s = ""
        while self.pos < len(self.text):
            c = self.peek()
            if c == q: self.advance(); break
            if c == '\\':
                self.advance(); e = self.advance()
                if e == 'u':
                    h = "".join([self.advance() for _ in range(4)])
                    s += chr(int(h, 16))
                elif e == 'U':
                    h = "".join([self.advance() for _ in range(8)])
                    s += chr(int(h, 16))
                else: s += {'n': '\n', 'r': '\r', 't': '\t'}.get(e, e)
            else: s += self.advance()
        return s
    def read_key(self):
        self.skip_ws(); c = self.peek()
        if c in ('"', "'"): return self.read_str(self.advance())
        k = ""
        while self.pos < len(self.text) and (self.text[self.pos].isalnum() or self.text[self.pos] in ('_', '-')): k += self.advance()
        return k
    def parse_value(self):
        self.skip_ws()
        if time.time() - self.start_time > 1: raise NolError("Timeout")
        self.depth += 1
        if self.depth > 100: raise NolError("Max depth exceeded")
        c = self.peek()
        if c == '{':
            self.advance(); o = {}
            while self.pos < len(self.text):
                self.skip_ws()
                if self.peek() == '}': self.advance(); break
                self.parse_pair(o); self.skip_ws()
                if self.peek() == ',': self.advance()
            res = o
        elif c == '[':
            self.advance(); a = []
            while self.pos < len(self.text):
                self.skip_ws()
                if self.peek() == ']': self.advance(); break
                a.append(self.parse_value()); self.skip_ws()
                if self.peek() == ',': self.advance()
            res = a
        elif self.nole and (c == '*' or c == '<'):
            t = self.advance()
            if t == '<':
                b = ""
                while self.peek() and self.peek() != '>': b += self.advance()
                if self.peek() == '>': self.advance()
                res = {"_coerce": {"type": b, "value": self.parse_value()}}
            else:
                b = ""
                while self.peek() and (self.peek().isalnum() or self.peek() in ('_', '-')): b += self.advance()
                res = "*" + b
        elif c in ('"', "'"): res = self.read_str(self.advance())
        elif c.isdigit() or c == '-':
            b = ""
            while self.peek() and (self.peek().isdigit() or self.peek() in ('.', '-', 'e', 'E', '+')): b += self.advance()
            res = float(b) if '.' in b or 'e' in b.lower() else int(b)
        else:
            b = ""
            while self.peek() and self.peek().isalpha(): b += self.advance()
            if b == "true": res = True
            elif b == "false": res = False
            elif b == "null": res = None
            else: raise NolError(f"Invalid value: {b}")
        self.depth -= 1; return res
    def parse_pair(self, obj):
        self.skip_ws()
        if self.nole and self.peek() == '&':
            self.advance(); n = self.read_key(); self.skip_ws()
            val = self.parse_value() if self.peek() == ':' and (self.advance() or True) else None
            if "_anchors" not in obj: obj["_anchors"] = []
            obj["_anchors"].append({"name": n, "value": val}); return
        is_m = False
        if self.nole and self.peek() == '<' and self.peek(1) == '<': self.advance(); self.advance(); is_m = True
        key = unicodedata.normalize('NFC', self.read_key()) if not is_m else "<<"
        self.skip_ws()
        if self.advance() != ':': raise NolError(f"Expected : for key {key}")
        val = self.parse_value()
        if is_m:
            if "<<" not in obj: obj["<<" ] = []
            obj["<<"].append(val)
        else:
            if key in obj: raise NolError(f"Duplicate key: {key}")
            obj[key] = val
    def parse_into(self, obj):
        while self.pos < len(self.text):
            self.skip_ws(); c = self.peek()
            if not c or c == '[': break
            self.parse_pair(obj)
    def parse(self):
        root = {}
        while self.pos < len(self.text):
            self.skip_ws()
            if not self.peek(): break
            if self.peek() == '[':
                self.advance(); is_a = False
                if self.peek() == '*': self.advance(); is_a = True
                path = ""
                while self.peek() and self.peek() != ']': path += self.advance()
                if self.peek() == ']': self.advance()
                parts = path.split('.'); curr = root
                for i, p in enumerate(parts):
                    last = (i == len(parts) - 1)
                    if p not in curr: curr[p] = [] if last and is_a else {}
                    if last:
                        if is_a:
                            entry = {}; curr[p].append(entry); self.parse_into(entry)
                        else: self.parse_into(curr[p])
                    else: curr = curr[p]
            else: self.parse_pair(root)
        return root

class Evaluator:
    def __init__(self, app_env=None):
        self.anchors = {}; self.app_env = set(app_env or []); self.doc_env = set(); self.root = None
    def evaluate(self, root):
        self.root = self.collect_meta(root)
        self.root = self.resolve_merges(self.root)
        self.root = self.resolve_env(self.root)
        import copy
        clone = copy.deepcopy(self.root)
        self.root = self.resolve_interp(self.root, clone)
        return self.resolve_coerce(self.root)
    def collect_meta(self, v):
        if isinstance(v, dict):
            if "_anchors" in v:
                ans = v.pop("_anchors")
                for a in ans:
                    name = a["name"]; val = self.collect_meta(a["value"]) if a["value"] is not None else v
                    self.anchors[name] = val
            if "_env" in v and isinstance(v["_env"], dict):
                for x in v["_env"].get("allowed", []): self.doc_env.add(x)
            for k in list(v.keys()): v[k] = self.collect_meta(v[k])
        elif isinstance(v, list): v = [self.collect_meta(x) for x in v]
        return v
    def resolve_merges(self, v, depth=0):
        if depth > 20: raise NolError("Max merge depth exceeded")
        if isinstance(v, dict):
            if "<<" in v:
                merges = v.pop("<<")
                if not isinstance(merges, list): merges = [merges]
                for m in merges:
                    rm = m
                    if isinstance(rm, str) and rm.startswith("*"):
                        an = rm[1:]; rm = self.anchors.get(an)
                    rm = self.resolve_merges(rm, depth + 1)
                    if isinstance(rm, dict):
                        for mk, mv in rm.items():
                            if not mk.startswith("_") and mk not in v: v[mk] = mv
            for k in list(v.keys()): v[k] = self.resolve_merges(v[k], depth)
        elif isinstance(v, list): v = [self.resolve_merges(x, depth) for x in v]
        return v
    def resolve_env(self, v):
        if isinstance(v, dict):
            if len(v) == 1 and "env" in v and isinstance(v["env"], str):
                var = v["env"]
                if not self.app_env and not self.doc_env or var in self.app_env or var in self.doc_env:
                    return os.environ.get(var, "")
            return {k: self.resolve_env(val) for k, val in v.items()}
        if isinstance(v, list): return [self.resolve_env(x) for x in v]
        return v
    def resolve_interp(self, v, root, depth=0):
        if depth > 50: raise NolError("Max interpolation depth exceeded")
        if isinstance(v, str) and "${" in v:
            def repl(m):
                path = m.group(1); curr = root
                for p in path.split('.'): curr = curr[p]
                vs = curr if isinstance(curr, str) else dump(curr, indent=2, root=True)
                if "${" in vs: vs = self.resolve_interp(vs, root, depth + 1)
                return vs
            return re.sub(r"\$\{([^}]+)\}", repl, v)
        if isinstance(v, dict): return {k: self.resolve_interp(val, root, depth) for k, val in v.items()}
        if isinstance(v, list): return [self.resolve_interp(x, root, depth) for x in v]
        return v
    def resolve_coerce(self, v):
        if isinstance(v, dict):
            if "_coerce" in v:
                c = v.pop("_coerce"); t = c["type"]; val = self.resolve_coerce(c["value"])
                s = val if isinstance(val, str) else dump(val, indent=2)
                if t == "int": return int(s)
                if t == "float": return float(s)
                if t == "bool": return s.lower() == "true"
                return s
            return {k: self.resolve_coerce(val) for k, val in v.items()}
        if isinstance(v, list): return [self.resolve_coerce(x) for x in v]
        return v

def parse(text, app_env=None):
    p = NolParser(text, True); root = p.parse()
    return Evaluator(app_env).evaluate(root)

def dump(v, indent=2, level=0, root=False):
    if v is None: return "null"
    if isinstance(v, bool): return "true" if v else "false"
    if isinstance(v, (int, float)):
        s = f"{v:.5f}" if isinstance(v, float) else str(v)
        if '.' in s: s = s.rstrip('0').rstrip('.')
        if '.' in s and s.endswith('.'): s += '0'
        return s
    if isinstance(v, str): return v if root else f'"{v}"'
    if isinstance(v, list): return "[" + ", ".join(dump(x, indent, level+1) for x in v) + "]"
    if isinstance(v, dict):
        if not root:
            if not v: return "{}"
            pad = " " * (level * indent); next_pad = " " * ((level + 1) * indent)
            items = []
            for k in sorted(v.keys()):
                if k.startswith("_"): continue
                items.append(f"\n{next_pad}{k}: {dump(v[k], indent, level+1)}")
            return "{" + ",".join(items) + "\n" + pad + "}"
        else:
            lines = []
            for k in sorted(v.keys()):
                if k.startswith("_"): continue
                lines.append(f"{k}: {dump(v[k], indent, level+1)}")
            return "\n".join(lines)

File: include/nole.js
const fs = require('fs');

class NolError extends Error {
    constructor(msg, line, col) {
        super(`${msg} at ${line}:${col}`);
        this.line = line; this.col = col;
    }
}

class NolParser {
    constructor(text, nole = false) {
        this.text = text; this.pos = 0; this.line = 1; this.col = 1;
        this.startTime = Date.now(); this.depth = 0; this.nole = nole;
    }
    peek(n = 0) { return this.text[this.pos + n] || null; }
    advance() {
        const c = this.peek(); this.pos++;
        if (c === '\n') { this.line++; this.col = 1; }
        else this.col++;
        return c;
    }
    skipWs() {
        while (this.pos < this.text.length) {
            const c = this.text[this.pos];
            if (/\s/.test(c)) this.advance();
            else if (c === '#') {
                this.advance();
                if (this.peek() === '#') {
                    this.advance();
                    while (this.pos < this.text.length - 1 && !(this.text[this.pos] === '#' && this.text[this.pos+1] === '#')) this.advance();
                    if (this.pos < this.text.length) { this.advance(); this.advance(); }
                } else {
                    while (this.pos < this.text.length && this.text[this.pos] !== '\n') this.advance();
                }
            } else break;
        }
    }
    readStr(q) {
        let s = "";
        while (this.pos < this.text.length) {
            const c = this.peek();
            if (c === q) { this.advance(); break; }
            if (c === '\\') {
                this.advance(); const e = this.advance();
                if (e === 'u') {
                    let h = this.text.slice(this.pos, this.pos + 4); this.pos += 4;
                    s += String.fromCharCode(parseInt(h, 16));
                } else if (e === 'U') {
                    let h = this.text.slice(this.pos, this.pos + 8); this.pos += 8;
                    s += String.fromCodePoint(parseInt(h, 16));
                } else s += {'n': '\n', 'r': '\r', 't': '\t'}[e] || e;
            } else s += this.advance();
        }
        return s;
    }
    readKey() {
        this.skipWs(); const c = this.peek();
        if (c === '"' || c === "'") return this.readStr(this.advance());
        let k = "";
        while (this.pos < this.text.length && /[a-zA-Z0-9_-]/.test(this.text[this.pos])) k += this.advance();
        return k;
    }
    parseValue() {
        this.skipWs();
        if (Date.now() - this.startTime > 1000) throw new NolError("Timeout", this.line, this.col);
        if (++this.depth > 100) throw new NolError("Max depth", this.line, this.col);
        const c = this.peek(); let res;
        if (c === '{') {
            this.advance(); const o = {};
            while (this.pos < this.text.length) {
                this.skipWs(); if (this.peek() === '}') { this.advance(); break; }
                this.parsePair(o); this.skipWs(); if (this.peek() === ',') this.advance();
            }
            res = o;
        } else if (c === '[') {
            this.advance(); const a = [];
            while (this.pos < this.text.length) {
                this.skipWs(); if (this.peek() === ']') { this.advance(); break; }
                a.push(this.parseValue()); this.skipWs(); if (this.peek() === ',') this.advance();
            }
            res = a;
        } else if (this.nole && (c === '*' || c === '<')) {
            const t = this.advance();
            if (t === '<') {
                let b = ""; while (this.peek() && this.peek() !== '>') b += this.advance();
                if (this.peek() === '>') this.advance();
                res = {"_coerce": {"type": b, "value": this.parseValue()}};
            } else {
                let b = ""; while (this.peek() && /[a-zA-Z0-9_-]/.test(this.peek())) b += this.advance();
                res = "*" + b;
            }
        } else if (c === '"' || c === "'") res = this.readStr(this.advance());
        else if (/[0-9-]/.test(c)) {
            let b = ""; while (this.peek() && /[0-9.eE+-]/.test(this.peek())) b += this.advance();
            res = b.includes('.') || /[eE]/.test(b) ? parseFloat(b) : parseInt(b);
        } else {
            let b = ""; while (this.peek() && /[a-z]/.test(this.peek())) b += this.advance();
            if (b === "true") res = true; else if (b === "false") res = false; else if (b === "null") res = null;
            else throw new NolError("Invalid value: " + b, this.line, this.col);
        }
        this.depth--; return res;
    }
    parsePair(obj) {
        this.skipWs();
        if (this.nole && this.peek() === '&') {
            this.advance(); const n = this.readKey(); this.skipWs();
            let val = null; if (this.peek() === ':') { this.advance(); val = this.parseValue(); }
            if (!obj._anchors) obj._anchors = [];
            obj._anchors.push({name: n, value: val}); return;
        }
        let isM = false; if (this.nole && this.peek() === '<' && this.peek(1) === '<') { this.advance(); this.advance(); isM = true; }
        const key = isM ? "<<" : this.readKey().normalize('NFC');
        this.skipWs(); if (this.advance() !== ':') throw new NolError("Expected :", this.line, this.col);
        const val = this.parseValue();
        if (isM) { if (!obj["<<"]) obj["<<"] = []; obj["<<"].push(val); }
        else {
            if (Object.prototype.hasOwnProperty.call(obj, key)) throw new Error("Duplicate key: " + key);
            obj[key] = val;
        }
    }
    parseInto(obj) {
        while (this.pos < this.text.length) {
            this.skipWs(); if (!this.peek() || this.peek() === '[') break;
            this.parsePair(obj);
        }
    }
    parse() {
        const root = {};
        while (this.pos < this.text.length) {
            this.skipWs(); if (!this.peek()) break;
            if (this.peek() === '[') {
                this.advance(); let isA = false; if (this.peek() === '*') { this.advance(); isA = true; }
                let path = ""; while (this.peek() && this.peek() !== ']') path += this.advance();
                if (this.peek() === ']') this.advance();
                const parts = path.split('.'); let curr = root;
                for (let i = 0; i < parts.length; i++) {
                    const p = parts[i], last = i === parts.length - 1;
                    if (!curr[p]) curr[p] = (last && isA) ? [] : {};
                    if (last) {
                        if (isA) { const entry = {}; curr[p].push(entry); this.parseInto(entry); }
                        else this.parseInto(curr[p]);
                    } else curr = curr[p];
                }
            } else this.parsePair(root);
        }
        return root;
    }
}

class Evaluator {
    constructor(appEnv = []) { this.anchors = {}; this.appEnv = new Set(appEnv); this.docEnv = new Set(); }
    evaluate(root) {
        this.root = this.collectMeta(root);
        this.root = this.resolveMerges(this.root);
        this.root = this.resolveEnv(this.root);
        const clone = JSON.parse(JSON.stringify(this.root));
        this.root = this.resolveInterp(this.root, clone);
        return this.resolveCoerce(this.root);
    }
    collectMeta(v) {
        if (v && typeof v === 'object' && !Array.isArray(v)) {
            if (v._anchors) {
                const ans = v._anchors; delete v._anchors;
                for (const a of ans) this.anchors[a.name] = this.collectMeta(a.value === null ? v : a.value);
            }
            if (v._env && typeof v._env === 'object' && Array.isArray(v._env.allowed)) {
                for (const x of v._env.allowed) this.docEnv.add(x);
            }
            for (const k in v) v[k] = this.collectMeta(v[k]);
        } else if (Array.isArray(v)) return v.map(x => this.collectMeta(x));
        return v;
    }
    resolveMerges(v, depth = 0) {
        if (depth > 20) throw new Error("Max merge depth");
        if (v && typeof v === 'object' && !Array.isArray(v)) {
            if (v["<<"]) {
                let merges = v["<<"]; delete v["<<"]; if (!Array.isArray(merges)) merges = [merges];
                for (let m of merges) {
                    let rm = m; if (typeof rm === 'string' && rm.startsWith("*")) rm = this.anchors[rm.slice(1)];
                    rm = this.resolveMerges(rm, depth + 1);
                    if (rm && typeof rm === 'object' && !Array.isArray(rm)) {
                        for (const mk in rm) if (!mk.startsWith("_") && !Object.prototype.hasOwnProperty.call(v, mk)) v[mk] = JSON.parse(JSON.stringify(rm[mk]));
                    }
                }
            }
            for (const k in v) v[k] = this.resolveMerges(v[k], depth);
        } else if (Array.isArray(v)) return v.map(x => this.resolveMerges(x, depth));
        return v;
    }
    resolveEnv(v) {
        if (v && typeof v === 'object' && !Array.isArray(v)) {
            const keys = Object.keys(v);
            if (keys.length === 1 && keys[0] === 'env' && typeof v.env === 'string') {
                if (this.appEnv.size === 0 && this.docEnv.size === 0 || this.appEnv.has(v.env) || this.docEnv.has(v.env)) return process.env[v.env] || "";
            }
            for (const k in v) v[k] = this.resolveEnv(v[k]);
        } else if (Array.isArray(v)) return v.map(x => this.resolveEnv(x));
        return v;
    }
    resolveInterp(v, root, depth = 0) {
        if (depth > 50) throw new Error("Max interp depth");
        if (typeof v === 'string' && v.includes("${")) {
            return v.replace(/\$\{([^}]+)\}/g, (_, path) => {
                let curr = root; for (const p of path.split('.')) curr = curr[p];
                let vs = typeof curr === 'string' ? curr : dump(curr, 2, 0, true);
                if (vs.includes("${")) vs = this.resolveInterp(vs, root, depth + 1);
                return vs;
            });
        }
        if (v && typeof v === 'object' && !Array.isArray(v)) { for (const k in v) v[k] = this.resolveInterp(v[k], root, depth); }
        else if (Array.isArray(v)) return v.map(x => this.resolveInterp(x, root, depth));
        return v;
    }
    resolveCoerce(v) {
        if (v && typeof v === 'object' && !Array.isArray(v)) {
            if (v._coerce) {
                const {type, value} = v._coerce; delete v._coerce;
                const val = this.resolveCoerce(value);
                const s = typeof val === 'string' ? val : dump(val, 2);
                if (type === "int") return parseInt(s); if (type === "float") return parseFloat(s);
                if (type === "bool") return s.toLowerCase() === "true"; return s;
            }
            for (const k in v) v[k] = this.resolveCoerce(v[k]);
        } else if (Array.isArray(v)) return v.map(x => this.resolveCoerce(x));
        return v;
    }
}

function parse(text, appEnv = []) {
    const p = new NolParser(text, true);
    return new Evaluator(appEnv).evaluate(p.parse());
}

function dump(v, indent = 2, level = 0, root = false) {
    if (v === null) return "null";
    if (typeof v === 'boolean') return v ? "true" : "false";
    if (typeof v === 'number') {
        let s = v.toFixed(5); while (s.endsWith('0')) s = s.slice(0, -1); if (s.endsWith('.')) s += '0';
        return s;
    }
    if (typeof v === 'string') return root ? v : `"${v}"`;
    if (Array.isArray(v)) return "[" + v.map(x => dump(x, indent, level + 1)).join(", ") + "]";
    const pad = " ".repeat(level * indent), nextPad = " ".repeat((level + 1) * indent);
    const keys = Object.keys(v).filter(k => !k.startsWith("_")).sort();
    if (root) return keys.map(k => `${k}: ${dump(v[k], indent, level + 1)}`).join("\n");
    if (keys.length === 0) return "{}";
    return "{" + keys.map(k => `\n${nextPad}${k}: ${dump(v[k], indent, level + 1)}`).join(",") + `\n${pad}}`;
}

module.exports = { parse, dump };
